### ISSUE TITLE
Load video frames

### ISSUE THREAD

**Issue by ltnghia** [facebookresearch/maskrcnn-benchmark#404]
## ❓ Questions and Help
I intend to implement Mask R-CNN for instance segmentation in videos.
I don't want to apply frame-by-frame but working for video blocks.
But maskrcnn-benchmark supports only loading images.
Could you please tell me how to load video frames as well as apply transformation to them?

**Comment by fmassa** [comment]
You'll need to first load the videos somehow, maybe using opencv.
From the individual frames, you can just apply the model for a sequence of frames at once (maybe as a batch?).

But extracting the frames of the video into images is still the easiest way to get around this. You could have a custom sampler that always returns adjacent frames for the same batch, so that it looks as if you had loaded it from a single video.

**Comment by ltnghia** [comment]
@fmassa: "extract the frames of the video into images", I'm doing in this way. But I don't know how to create _dataloader_ to load and transform many images at the same time. Can you show me some ideas?

**Comment by fmassa** [comment]
A few questions:
- do you have 3d convolutions in your model, or just 2d convolutions?
- the only random augmentation that we currently do is flip

If you want to train with 2d convolution, the easy thing to do is to just disable flip for now.

**Comment by fmassa** [comment]
If you have prior knowledge of PyTorch, the images are all loaded in the [`DataLoader`](https://pytorch.org/docs/stable/data.html?highlight=dataloader#torch.utils.data.DataLoader), with the order specified by the [`Sampler`](https://pytorch.org/docs/stable/data.html?highlight=sampler#torch.utils.data.Sampler).

I can point more details in the `maskrcnn-benchmark` implementation if you wish, but I'd first recommend you get familiar with those concepts.

**Comment by ltnghia** [comment]
Please show me more details in the maskrcnn-benchmark implementation.

**Comment by fmassa** [comment]
Here is where the samplers are defined
https://github.com/facebookresearch/maskrcnn-benchmark/blob/519e8dd4402ed13365948523960ef24fb5dead92/maskrcnn_benchmark/data/build.py#L59-L66
you can create a new sampler which will output images that come sequentially from a video sequence.
For example:
```python
class SequenceFrameSampler(Sampler):
    def __init__(self, dataset_video_list):
        # this is a list of list of names.
        # the first level correspond to a video, and the second
        # one to the names of the frames in the video
        self.dataset_video_list = dataset_video_list
    
    def __iter__(self):
        # approximation for now, only take the first 8 frames per video
        video_order = torch.randperm(self.dataset_video_list).tolist()
        for i in video_order:
            for frames in self.dataset_video_list[video_order[i]][:8]:
                yield frames

    def __len__(self):
        return len(self.dataset_video_list) * 8
```

This is an untested snippet, but the idea is there.

**Comment by ltnghia** [comment]
I try implementing it.
Thank you so much.

**Comment by ltnghia** [comment]
I implemented _SequenceFrameSampler_ as your idea, and in _trainer_:
```
window_size = 8
for iteration, (images, targets, _) in enumerate(data_loader, start_iter):
     if (iteration % window_size) == 0:
            data['image'] = []
            data['target'] = []
     data['image'].append(images)
     data['target'].append(targets)
     if (iteration % window_size) != (window_size - 1):
            continue
     # further processing for an 8-frames block
```
I can get blocks of 8 frames. But in this way, I use only batch size 1 and 1 GPU. How can I implement to increase batch size for multi GPUs?

**Comment by fmassa** [comment]
You can follow a similar approach, but you'll need to handle videos of different sizes.
I'd recommend having a look at `DistributedSampler`, this will be helpful

**Comment by ltnghia** [comment]
Can I feed the whole block and batch into each GPU? 
For example, I want to use batch size 4 of 8-frame blocks with 2 GPUs. But the current implementation of _SequenceFrameSampler_ supports only batch size 1. I have to solve this point before modifying _DistributedSampler_, because it doesn't solve the problem:

It should be:
[batch0-1]:   0,   1,   2,  3,   4,   5,    6,  7
[batch0-2]:   8,   9, 10, 11, 12, 13, 14, 15
[batch0-3]: 16, 17, 18, 19, 20, 21, 22, 23
[batch0-4]: 24, 25, 26, 27, 28, 29, 30, 31

But current implemented _SequenceFrameSampler_ gives:
[batch0-1]:   0,   4,   8,  12,  16,  20,  24,  28
[batch0-2]:   1,   5,   9,  13,  17,  21,  25,  29
[batch0-3]:   2,   6, 10,  14,  18,  22,  26,  30
[batch0-4]:   3,   7, 11,  15,  19,  23,  27,  31

**Comment by fmassa** [comment]
@ltnghia Sorry, I don't quite follow. Just make sure that the Sampler returns the values that you want. You can test this locally in IPython.
Once the sampler is returning the correct indices, you can look into adding support for multi-gpu via `DistributedSampler`.

**Comment by ltnghia** [comment]
I re-implemented these modules.
Great thanks for your help.

**Comment by fmassa** [comment]
Great, let me know if you have further questions, or if you think there might be a small change in the codebase that could be improved with a PR.
